core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java

```diff
@@ -0,0 +1,93 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.xcontent.XContentHelper.toXContent;
+
+public class ShardSearchFailureTests extends ESTestCase {
+
+    public static ShardSearchFailure createTestItem() {
+        String randomMessage = randomAsciiOfLengthBetween(3, 20);
+        Exception ex = new ParsingException(0, 0, randomMessage , null);
+        String nodeId = randomAsciiOfLengthBetween(5, 10);
+        String indexName = randomAsciiOfLengthBetween(5, 10);
+        String indexUuid = randomAsciiOfLengthBetween(5, 10);
+        int shardId = randomInt();
+        return new ShardSearchFailure(ex,
+                new SearchShardTarget(nodeId, new ShardId(new Index(indexName, indexUuid), shardId)));
+    }
+
+    public void testFromXContent() throws IOException {
+        ShardSearchFailure response = createTestItem();
+        XContentType xcontentType = randomFrom(XContentType.values());
+        XContentBuilder builder = XContentFactory.contentBuilder(xcontentType);
+        builder.startObject();
+        builder = response.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        builder.endObject();
+
+        XContentParser parser = createParser(builder);
+        parser.nextToken(); // jump to first START_OBJECT
+        ShardSearchFailure parsed = ShardSearchFailure.fromXContent(parser);
+
+        assertEquals(response.index(), parsed.index());
+        assertEquals(response.shard().nodeId(), parsed.shard().nodeId());
+        assertEquals(response.shardId(), parsed.shardId());
+
+        // we cannot compare the cause, because it will be wrapped in an outer ElasticSearchException
+        // best effort: try to check that the original message appears somewhere in the rendered xContent
+        String originalMsg = response.getCause().getMessage();
+        assertTrue(parsed.getCause().getMessage().contains(originalMsg));
```

Comment:
It can be complicated to rebuild the object, how about doing something like:
```
assertEquals(parsed.getCause().getMessage(), "Elasticsearch exception [type=parse_exception, reason=" + originalMsg +"]");
```